File: lib/to_mdx/parsers/open_document/text/link.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Text
      class Link < Base
        attr_reader :href, :style_name

        def initialize(element = nil, context = {})
          super(element, context)
          @href = element&.attributes&.[]("xlink:href")
          @style_name = context[:styles].use(element&.attributes&.[]("text:style-name"))
        end

        def empty?
          !@href || @href.empty?
        end

        def to_mdx
          alt_text = contentful_children.map(&:to_mdx).join
          if @style_name.empty?
            "<a href=\"#{@href}\">#{alt_text}</a>"
          else
            "<a href=\"#{@href}\" className=\"#{@style_name}\">#{alt_text}</a>"
          end
        end

        def to_s
          alt_text = contentful_children.map(&:to_s).join
          "[#{alt_text}](#{@href})"
        end

        register_for "text:a"
      end
    end
  end
end

File: src/parsers/open-document/text/link.ts
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class Link extends BaseElement {
  private href: string;
  private styleName: Style | undefined = undefined;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.href = this.attr("xlink:href") || "";
    this.styleName = context.styles.use(this.attr("text:style-name"));
  }

  empty(): boolean {
    return !this.href || this.href.length === 0;
  }

  toMdx(): string {
    const altText = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join("");

    if (this.styleName?.empty()) {
      return `<a href="${this.href}">${altText}</a>`;
    }

    return `<a href="${this.href}" className="${this.styleName}">${altText}</a>`;
  }

  toString(): string {
    const altText = this.contentfulChildren()
      .map((child) => child.toString())
      .join("");
    return `[${altText}](${this.href})`;
  }

  static {
    BaseElement.registerFor.call(this, "text:a");
  }
}
